rust/agents/validator/src/validator.rs

```diff
@@ -74,17 +103,139 @@ impl BaseAgent for Validator {
 
     #[allow(clippy::async_yields_async)]
     async fn run(&self) -> Instrumented<JoinHandle<Result<()>>> {
-        let submit = ValidatorSubmitter::new(
+        self.announce().await.expect("Failed to announce validator");
+
+        let mut tasks = vec![];
+
+        tasks.push(self.run_message_sync().await);
+        for checkpoint_sync_task in self.run_checkpoint_submitters().await {
+            tasks.push(checkpoint_sync_task);
+        }
+
+        run_all(tasks)
+    }
+}
+
+impl Validator {
+    async fn run_message_sync(&self) -> Instrumented<JoinHandle<eyre::Result<()>>> {
+        let index_settings = self.as_ref().settings.chains[self.origin_chain.name()]
+            .index
+            .clone();
+        let contract_sync = self.message_sync.clone();
+        let cursor = contract_sync
+            .forward_backward_message_sync_cursor(index_settings.chunk_size)
+            .await;
+        tokio::spawn(async move {
+            contract_sync
+                .clone()
+                .sync("dispatched_messages", cursor)
+                .await
+        })
+        .instrument(info_span!("MailboxMessageSyncer"))
+    }
+
+    async fn run_checkpoint_submitters(&self) -> Vec<Instrumented<JoinHandle<eyre::Result<()>>>> {
+        let submitter = ValidatorSubmitter::new(
             self.interval,
             self.reorg_period,
             self.mailbox.clone(),
-            self.validator_announce.clone(),
             self.signer.clone(),
             self.checkpoint_syncer.clone(),
+            self.db.clone(),
             ValidatorSubmitterMetrics::new(&self.core.metrics, &self.origin_chain),
         );
 
-        run_all(vec![submit.spawn()])
+        let empty_tree = IncrementalMerkle::default();
+        let lag = NonZeroU64::new(self.reorg_period);
+        let tip_tree = self
+            .mailbox
+            .tree(lag)
+            .await
+            .expect("failed to get mailbox tree");
+        let backfill_target = submitter.checkpoint(&tip_tree);
```

Comment:
Should we pull the checkpoint via Mailbox.checkpoint? This way safety relies less on the incremental merkle tree implementation in the agent